### PR TITLE
Issue 272: Multiple Message CommitBatch Update and No Offset Result Conversion

### PR DESCRIPTION
**Pull request by dbcfd**
https://github.com/akka/reactive-kafka/issues/272

Provide method to update commitbatch with multiple values (e.g. when grouping kafka messages), and provide a version of the KafkaConsumerActor that does not convert the commit result to a map.

### CODE REVIEW

File: core/src/main/scala/akka/kafka/internal/ConsumerStage.scala [hunk 1]
```diff
@@ -153,12 +153,11 @@ private[kafka] object ConsumerStage {
...
       extends CommittableOffsetBatch {
 
-    override def updated(committableOffset: CommittableOffset): CommittableOffsetBatch = {
-      val partitionOffset = committableOffset.partitionOffset
-      val key = partitionOffset.key
-
-      val newOffsets = offsets.updated(key, committableOffset.partitionOffset.offset)
-
+    private def updatedStages(
+      existing: Map[String, Committer],
```
Comment (patriknw) [inline core/src/main/scala/akka/kafka/internal/ConsumerStage.scala]: this is only used with `updatedStages(stages` so the `existing` parameter is not needed (`stages` can be used)

File: core/src/main/scala/akka/kafka/internal/ConsumerStage.scala [hunk 2]
```diff
@@ -167,14 +166,37 @@ private[kafka] object ConsumerStage {
...
+
+      val newOffsets = offsets.updated(key, committableOffset.partitionOffset.offset)
+
+      new CommittableOffsetBatchImpl(newOffsets, updatedStages(stages, key, committableOffset))
+    }
+
+    override def updated(committableOffsets: Seq[CommittableOffset]): CommittableOffsetBatch = {
+      val (newStages, newOffsets) = committableOffsets.groupBy(_.partitionOffset.key).foldLeft((stages, offsets)) {
+        case ((aggStages, aggOffsets), (key, committableOffsetsForKey)) =>
+          val commitToUse = committableOffsetsForKey.sortBy(_.partitionOffset.offset).head
```
Comment (patriknw) [inline core/src/main/scala/akka/kafka/internal/ConsumerStage.scala]: this is picking the smallest offset, what is the reason for that?

File: core/src/main/scala/akka/kafka/internal/ConsumerStage.scala [hunk 3]
```diff
@@ -167,14 +166,37 @@ private[kafka] object ConsumerStage {
...
+          val partitionOffset = commitToUse.partitionOffset
+          val key = partitionOffset.key
+          val offset = commitToUse.partitionOffset.offset
+
+          val offsetToUpdateTo = Math.max(aggOffsets.getOrElse(key, offset), offset)
```
Comment (patriknw) [inline core/src/main/scala/akka/kafka/internal/ConsumerStage.scala]: but here it's only using that offset if it's greater than existing in the batch
I'm not sure I understand what this method is intended to do.

Comment (patriknw) [inline core/src/main/scala/akka/kafka/internal/ConsumerStage.scala]: ScalaDoc describing semantics will be needed.

File: core/src/main/scala/akka/kafka/internal/ConsumerStage.scala [hunk 4]
```diff
@@ -97,13 +97,13 @@ private[kafka] object ConsumerStage {
       val offsets = Map(
         new TopicPartition(offset.key.topic, offset.key.partition) -> (offset.offset + 1)
       )
-      (ref ? Commit(offsets)).mapTo[Committed].map(_ => Done)
+      (ref ? Internal.Commit(offsets)).mapTo[Internal.CompletedCommit.type].map(_ => Done)
```
Comment (patriknw) [inline core/src/main/scala/akka/kafka/internal/ConsumerStage.scala]: how can we know that it's `CompletedCommit` and not `OffsetCommitResult`?

File: core/src/main/scala/akka/kafka/KafkaConsumerActor.scala [hunk 5]
```diff
@@ -7,21 +7,36 @@ package akka.kafka
...
 
 object KafkaConsumerActor {
   case class StoppingException() extends RuntimeException("Kafka consumer is stopping")
-  def props[K, V](settings: ConsumerSettings[K, V]): Props =
-    Props(new KafkaConsumerActor(settings)).withDispatcher(settings.dispatcher)
+
+  def propsWithCommitResult[K, V](
+    settings: ConsumerSettings[K, V]
+  ): Props = {
+    Props(new KafkaConsumerActor(settings, m => Internal.OffsetCommitResult(m.asScala.toMap))).withDispatcher(settings.dispatcher)
```
Comment (patriknw) [inline core/src/main/scala/akka/kafka/KafkaConsumerActor.scala]: I don't understand this, what is the purpose of this `OffsetCommitResult`? I don't see it used anywhere.

File: core/src/main/scala/akka/kafka/ConsumerMessage.scala [hunk 6]
```diff
@@ -90,6 +90,13 @@ object ConsumerMessage {
     def updated(offset: CommittableOffset): CommittableOffsetBatch
 
     /**
+     * Add/overwite an offset position for the given groupId, topic, partition.
+     * @param offsets Sequence of offsets to add/overwrite offset position for the given groupId, topic, partition.
+     * @return
+     */
+    def updated(offsets: Seq[CommittableOffset]): CommittableOffsetBatch
```
Comment (dbcfd) [inline core/src/main/scala/akka/kafka/ConsumerMessage.scala]: Will add that and the corresponding java api.

File: core/src/main/scala/akka/kafka/internal/ConsumerStage.scala (hunk 4 above)
Comment (dbcfd) [inline core/src/main/scala/akka/kafka/internal/ConsumerStage.scala]: Further discussion on your comment about use.

File: core/src/main/scala/akka/kafka/internal/ConsumerStage.scala (hunk 1 above)
Comment (dbcfd) [inline core/src/main/scala/akka/kafka/internal/ConsumerStage.scala]: Think I had originally meant for this to be an aggregation and not use stages, but it makes more sense just to use stages as implemented.

File: core/src/main/scala/akka/kafka/internal/ConsumerStage.scala (hunk 2 above)
Comment (dbcfd) [inline core/src/main/scala/akka/kafka/internal/ConsumerStage.scala]: Error, should be last for largest offset. I would like to add additional tests around this functionality. Best to do a test similar to https://github.com/akka/reactive-kafka/blob/master/core/src/test/scala/akka/kafka/scaladsl/IntegrationSpec.scala#L140?

File: core/src/main/scala/akka/kafka/internal/ConsumerStage.scala (hunk 3 above)
Comment (dbcfd) [inline core/src/main/scala/akka/kafka/internal/ConsumerStage.scala]: Meant to select largest offset both above and here, which should be the same behavior as a single offset.

File: core/src/main/scala/akka/kafka/KafkaConsumerActor.scala (hunk 5 above)
Comment (dbcfd) [inline core/src/main/scala/akka/kafka/KafkaConsumerActor.scala]: I had originally added this to handle cases where the commit result is used, but after adding it, found that no users of this actor seem to use the commit result. Am I missing the use of the commits somewhere? Does it make sense to expose the committed offsets in the api (e.g. commitableSourceWithOffsets) or should this actor just not return the committed offsets?

**Comment by ennru** [comment]
I looked into this PR now and can see a slight chance that the current code in certain situations could overwrite a higher commit offset with a lower one. So the logic to take the max of the current value in the offsets map and the incoming one could make sense.
For the rest of the PR I don't get enough motivation in this issue description and the code has moved forward a lot since you opened it.
I'll close both the PR and this issue, please open a new one if you have any updates on this.

**Pull request closed without merging**